File: 04_python_codes/vwo_login_selenium.py
import os
import random
import re
import string
from pathlib import Path
from typing import Optional

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException


def generate_random_email(domain: str = "example.com") -> str:
    local = "user_" + "".join(random.choice(string.ascii_lowercase + string.digits) for _ in range(10))
    return f"{local}@{domain}"


def ensure_directory(path: Path) -> None:
    if not path.parent.exists():
        path.parent.mkdir(parents=True, exist_ok=True)


def find_first_present(driver: webdriver.Chrome, selectors: list[str]) -> Optional[tuple[str, str]]:
    for selector in selectors:
        try:
            driver.find_element(By.CSS_SELECTOR, selector)
            return ("css", selector)
        except NoSuchElementException:
            continue
    return None


def main() -> None:
    screenshot_path = Path("c:/MCP_Agents/02_selenium_mcp/screenshots/vwo_login_result.png")
    ensure_directory(screenshot_path)

    options = webdriver.ChromeOptions()
    # Explicitly ensure non-headless for visibility
    # (By default Chrome opens with a UI; we avoid adding any --headless flags.)

    driver = webdriver.Chrome(options=options)
    driver.maximize_window()

    wait = WebDriverWait(driver, 15)

    try:
        driver.get("http://app.vwo.com")

        email_selectors = [
            "input[type='email']",
            "input[name*='email' i]",
            "input[id*='email' i]",
            "input[placeholder*='email' i]",
        ]
        password_selectors = [
            "input[type='password']",
            "input[name*='pass' i]",
            "input[id*='pass' i]",
            "input[placeholder*='password' i]",
        ]
        submit_selectors = [
            "button[type='submit']",
            "input[type='submit']",
            "button:where(:is(:-internal-has-text('Sign in'), :-internal-has-text('Log in'), :-internal-has-text('Login')))",
            "button",
        ]

        # Find and fill email
        email_locator = find_first_present(driver, email_selectors)
        if not email_locator:
            raise RuntimeError("Email input not found")
        email_element = driver.find_element(By.CSS_SELECTOR, email_locator[1])
        email_element.clear()
        email_element.send_keys(generate_random_email())

        # Find and fill password
        password_locator = find_first_present(driver, password_selectors)
        if not password_locator:
            raise RuntimeError("Password input not found")
        password_element = driver.find_element(By.CSS_SELECTOR, password_locator[1])
        password_element.clear()
        password_element.send_keys("Pwd_" + "".join(random.choice(string.ascii_letters + string.digits) for _ in range(12)))

        # Click submit
        submit_locator = find_first_present(driver, submit_selectors)
        if submit_locator:
            driver.find_element(By.CSS_SELECTOR, submit_locator[1]).click()
        else:
            password_element.send_keys(Keys.ENTER)

        # Wait for a likely error message to appear
        error_text: Optional[str] = None
        try:
            # Try ARIA role alert
            alert = wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, "[role='alert']")))
            error_text = alert.text.strip()
        except TimeoutException:
            # Try common error keywords in text nodes
            try:
                # Broad text match using XPath contains with case-insensitive translate
                error_node = wait.until(
                    EC.visibility_of_element_located(
                        (
                            By.XPATH,
                            "//*[contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'invalid') or "
                            "contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'incorrect') or "
                            "contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'wrong') or "
                            "contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'error') or "
                            "contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'failed')]",
                        )
                    )
                )
                error_text = error_node.text.strip()
            except TimeoutException:
                error_text = None

        # Take screenshot
        driver.save_screenshot(str(screenshot_path))

        if error_text:
            print(f"Observed error message: {error_text}")
        else:
            print("No explicit error message detected within timeout.")

    finally:
        driver.quit()


if __name__ == "__main__":
    main()

File: 02_selenium_mcp/VwoLoginSelenium.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.List;
import java.util.Locale;
import java.util.NoSuchElementException;
import java.util.Random;
import java.util.concurrent.TimeoutException;

import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.io.FileHandler;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public class VwoLoginSelenium {

    private static String randomEmail() {
        String chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        StringBuilder sb = new StringBuilder("user_");
        Random r = new Random();
        for (int i = 0; i < 10; i++) sb.append(chars.charAt(r.nextInt(chars.length())));
        return sb.append("@example.com").toString();
    }

    private static void ensureDirectory(Path filePath) throws IOException {
        Path parent = filePath.getParent();
        if (parent != null && !Files.exists(parent)) {
            Files.createDirectories(parent);
        }
    }

    private static WebElement firstPresent(WebDriver driver, List<By> bys) {
        for (By by : bys) {
            try {
                WebElement el = driver.findElement(by);
                if (el != null) return el;
            } catch (NoSuchElementException ignored) {}
        }
        return null;
    }

    public static void main(String[] args) throws IOException {
        Path screenshot = Paths.get("c:/MCP_Agents/02_selenium_mcp/screenshots/vwo_login_result_java.png");
        ensureDirectory(screenshot);

        ChromeOptions options = new ChromeOptions();
        // Non-headless by default
        WebDriver driver = new ChromeDriver(options);
        driver.manage().window().maximize();

        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(15));

        try {
            driver.get("http://app.vwo.com");

            WebElement email = firstPresent(driver, List.of(
                By.cssSelector("input[type='email']"),
                By.cssSelector("input[name*='email' i]"),
                By.cssSelector("input[id*='email' i]"),
                By.cssSelector("input[placeholder*='email' i]")
            ));
            if (email == null) throw new RuntimeException("Email input not found");
            email.clear();
            email.sendKeys(randomEmail());

            WebElement password = firstPresent(driver, List.of(
                By.cssSelector("input[type='password']"),
                By.cssSelector("input[name*='pass' i]"),
                By.cssSelector("input[id*='pass' i]"),
                By.cssSelector("input[placeholder*='password' i]")
            ));
            if (password == null) throw new RuntimeException("Password input not found");
            password.clear();
            password.sendKeys("Pwd_" + java.util.UUID.randomUUID());

            WebElement submit = firstPresent(driver, List.of(
                By.cssSelector("button[type='submit']"),
                By.cssSelector("input[type='submit']"),
                By.cssSelector("button")
            ));
            if (submit != null) submit.click(); else password.sendKeys(Keys.ENTER);

            String errorText = null;
            try {
                WebElement alert = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[role='alert']")));
                errorText = alert.getText().trim();
            } catch (TimeoutException ex) {
                try {
                    WebElement generic = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath(
                        "//*[contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'invalid') or " +
                        "contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'incorrect') or " +
                        "contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'wrong') or " +
                        "contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'error') or " +
                        "contains(translate(normalize-space(string(.)), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'failed')]"
                    )));
                    errorText = generic.getText().trim();
                } catch (TimeoutException ignored) {}
            }

            File src = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
            FileHandler.copy(src, screenshot.toFile());

            if (errorText != null && !errorText.isEmpty()) {
                System.out.println("Observed error message: " + errorText);
            } else {
                System.out.println("No explicit error message detected within timeout.");
            }

        } finally {
            driver.quit();
        }
    }
}
